lib/sdotypes.c

```diff
@@ -3569,6 +3658,609 @@ bool sdo_encrypted_packet_windup(sdow_t *sdow, int type, sdo_iv_t *iv)
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read RVIP\n");
+		return false;
+	}
+
+	size_t rvdns_length = 0;
+	if (!sdor_string_length(sdor, &rvdns_length) || rvdns_length == 0) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read RVDNS length\n");
+		return false;
+	}
+	rvto2addr_entry->rvdns = sdo_string_alloc_size(rvdns_length);
+	if (!rvto2addr_entry->rvdns) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to alloc RVDNS\n");
+		return false;
+	}
+	
+	if (!sdor_text_string(sdor, rvto2addr_entry->rvdns->bytes, rvto2addr_entry->rvdns->byte_sz)) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read RVDNS\n");
+		return false;
+	}
+
+	rvto2addr_entry->rvport = -1;
+	if (!sdor_signed_int(sdor, &rvto2addr_entry->rvport) ||
+		rvto2addr_entry->rvport == -1) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read RVPort\n");
+		return false;
+	}
+
+	rvto2addr_entry->rvprotocol = -1;
+	if (!sdor_signed_int(sdor, &rvto2addr_entry->rvprotocol) ||
+		rvto2addr_entry->rvprotocol == -1) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read RVProtocol\n");
+		return false;
+	}
+
+	if (!sdor_end_array(sdor)) {
+		LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read end array\n");
+		goto end;
+	}
+	return true;
+end:
+	fdo_rvto2addr_entry_free(rvto2addr_entry);
+	return false;
+}
+
+/**
+ * Read RVTO2Addr into the given fdo_rvto2addr_t object.
+ * Memory allocation for the internal elements of fdo_rvto2addr_t object
+ * will be done in this method.
+ * However, the memory must be allocated for fdo_rvto2addr_y_t object and
+ * given to this method.
+ * [
+ * +RVTO2AddrEntry 		// one or more RVTO2AddrEntry
+ * ]
+ * Return true, if read was a success. False otherwise.
+ */
+bool fdo_rvto2addr_read(sdor_t *sdor, fdo_rvto2addr_t *rvto2addr) {
+	size_t num_rvto2addr_items = 0;
+	if (!sdor_array_length(sdor, &num_rvto2addr_items) || num_rvto2addr_items == 0) {
+		LOG(LOG_ERROR, "RVTO2Addr: Failed to read/Invalid array length\n");
+		return false;
+	}
+
+	if (!sdor_start_array(sdor)) {
+		LOG(LOG_ERROR, "RVTO2Addr: Failed to read/Invalid array length\n");
+		return false;
+	}
+
+	LOG(LOG_DEBUG, "RVTO2Addr: There are %zu RVTO2AddrEntry(s)\n", num_rvto2addr_items);
+
+	rvto2addr->num_rvto2addr = num_rvto2addr_items;
+	rvto2addr->rv_to2addr_entry = sdo_alloc(sizeof(fdo_rvto2addr_entry_t));
+	if (!rvto2addr->rv_to2addr_entry) {
+		LOG(LOG_ERROR, "RVTO2Addr: Failed to alloc RVTO2AddrEntry\n");
+		return false;	
+	}
+	fdo_rvto2addr_entry_t *entry = rvto2addr->rv_to2addr_entry;
+	size_t i = 0;
+	for (;;) {
+
+		i++;
+		if (!fdo_rvto2addr_entry_read(sdor, entry)) {
+			LOG(LOG_ERROR, "RVTO2Addr: Failed to read RVTO2AddrEntry\n");
+			goto end;
+		}
+		if (i < num_rvto2addr_items) {
+			entry->next = sdo_alloc(sizeof(fdo_rvto2addr_entry_t));
+			if (!entry->next) {
+				LOG(LOG_ERROR, "RVTO2AddrEntry: Failed to read/Invalid array length\n");
+				goto end;
+			}
+			entry = (fdo_rvto2addr_entry_t *) entry->next;
```

Comment:
Typecasting not required as `entry->next` is of the same type.

Reply:
Done.